dotnet-authserver/tests/TeacherIdentity.AuthServer.Tests/Services/UserImport/UserImportProcessorTests.cs

```diff
@@ -0,0 +1,1057 @@
+using System.Text;
+using System.Text.Json;
+using Microsoft.EntityFrameworkCore;
+using TeacherIdentity.AuthServer.Events;
+using TeacherIdentity.AuthServer.Models;
+using TeacherIdentity.AuthServer.Services.UserImport;
+using User = TeacherIdentity.AuthServer.Models.User;
+
+namespace TeacherIdentity.AuthServer.Tests.Services;
+
+public class UserImportProcessorTests : IClassFixture<DbFixture>
+{
+    private readonly DbFixture _dbFixture;
+
+    public UserImportProcessorTests(DbFixture dbFixture)
+    {
+        _dbFixture = dbFixture;
+    }
+
+    [Fact]
+    public async Task Process_WithValidCsvData_InsertsUserAndEventAndUserImportJobRowIntoDatabase()
+    {
+        // Arrange
+        var dbContext = _dbFixture.GetDbContext();
+        var userImportStorageService = new Mock<IUserImportStorageService>();
+        var clock = new TestClock();
+        var logger = new Mock<ILogger<UserImportProcessor>>();
+        var userImportJobId = Guid.NewGuid();
+        var storedFilename = $"{userImportJobId}.csv";
+        var csvContent = new StringBuilder();
+        csvContent.AppendLine("ID,EMAIL_ADDRESS,FIRST_NAME,LAST_NAME,DATE_OF_BIRTH");
+        csvContent.AppendLine("UserImportProcessorTests1,UserImportProcessorTests1@email.com,joe,bloggs,05021970");
+        using var csvStream = new MemoryStream(Encoding.UTF8.GetBytes(csvContent.ToString()));
+
+        await _dbFixture.TestData.WithDbContext(async dbContext =>
+        {
+            dbContext.UserImportJobs.Add(new UserImportJob()
+            {
+                UserImportJobId = userImportJobId,
+                StoredFilename = storedFilename,
+                OriginalFilename = "my-user-import.csv",
+                UserImportJobStatus = UserImportJobStatus.New,
+                Uploaded = clock.UtcNow
+            });
+
+            await dbContext.SaveChangesAsync();
+        });
+
+        userImportStorageService.Setup(s => s.OpenReadStream(storedFilename))
+            .ReturnsAsync(csvStream)
+            .Verifiable();
+
+        // Act
+        var userImportProcessor = new UserImportProcessor(
+            dbContext,
+            userImportStorageService.Object,
+            clock,
+            logger.Object);
+        await userImportProcessor.Process(userImportJobId);
+
+        // Assert
+        userImportStorageService
+            .Verify();
+
+        await _dbFixture.TestData.WithDbContext(async dbContext =>
+        {
+            var user = await dbContext.Users.SingleOrDefaultAsync(r => r.EmailAddress == "UserImportProcessorTests1@email.com");
+            Assert.NotNull(user);
+
+            var events = await dbContext.Events.Where(e => e.EventName == "UserImportedEvent").ToListAsync();
+            var userImportedEventCount = events
+                .Select(e => JsonSerializer.Deserialize<UserImportedEvent>(e.Payload))
+                .Count(i => i!.UserImportJobId == userImportJobId && i.User.UserId == user.UserId);
```

Comment:
Using `Assert.Collection` on the events might be a little cleaner than count

Reply:
I've done this slighlty differently in a single test for all theories